File: rag-quickstart/python/embedding.py
"""Real embedding providers selectable by config: openai, azure-openai, ollama.

Each provider implements EmbeddingProvider.embed(text) -> list[float] by calling
the provider's real HTTP embeddings API. The provider is chosen from
embedding.provider via create_provider().
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

import requests

from config_loader import EmbeddingConfig

AZURE_API_VERSION = "2024-02-01"
HTTP_TIMEOUT_SECONDS = 60


@runtime_checkable
class EmbeddingProvider(Protocol):
    def embed(self, text: str) -> list[float]:
        """Return the embedding vector for ``text``."""
        ...


class OpenAIProvider:
    """OpenAI embeddings API (https://api.openai.com/v1/embeddings)."""

    def __init__(self, config: EmbeddingConfig):
        self._config = config

    def embed(self, text: str) -> list[float]:
        response = requests.post(
            "https://api.openai.com/v1/embeddings",
            headers={"Authorization": f"Bearer {self._config.api_key}"},
            json={"model": self._config.model, "input": text},
            timeout=HTTP_TIMEOUT_SECONDS,
        )
        response.raise_for_status()
        return response.json()["data"][0]["embedding"]


class AzureOpenAIProvider:
    """Azure OpenAI embeddings (deployment-scoped URL, ``api-key`` header)."""

    def __init__(self, config: EmbeddingConfig):
        self._config = config

    def embed(self, text: str) -> list[float]:
        endpoint = self._config.endpoint.rstrip("/")
        url = (
            f"{endpoint}/openai/deployments/"
            f"{self._config.deployment_name}/embeddings"
        )
        response = requests.post(
            url,
            params={"api-version": AZURE_API_VERSION},
            headers={"api-key": self._config.api_key},
            json={"input": text},
            timeout=HTTP_TIMEOUT_SECONDS,
        )
        response.raise_for_status()
        return response.json()["data"][0]["embedding"]


class OllamaProvider:
    """Local Ollama embeddings (no API key, POST /api/embeddings)."""

    def __init__(self, config: EmbeddingConfig):
        self._config = config

    def embed(self, text: str) -> list[float]:
        endpoint = self._config.endpoint.rstrip("/")
        response = requests.post(
            f"{endpoint}/api/embeddings",
            json={"model": self._config.model, "prompt": text},
            timeout=HTTP_TIMEOUT_SECONDS,
        )
        response.raise_for_status()
        return response.json()["embedding"]


PROVIDER_REGISTRY: dict[str, type] = {
    "openai": OpenAIProvider,
    "azure-openai": AzureOpenAIProvider,
    "ollama": OllamaProvider,
}


def create_provider(config: EmbeddingConfig) -> EmbeddingProvider:
    """Return the provider implementation selected by ``config.provider``."""
    provider_class = PROVIDER_REGISTRY.get(config.provider)
    if provider_class is None:
        raise ValueError(
            f"Unknown embedding provider: {config.provider!r}. "
            f"Supported: {', '.join(sorted(PROVIDER_REGISTRY))}."
        )
    return provider_class(config)

File: rag-quickstart/typescript/src/embedding.ts
/**
 * Real embedding providers selectable by config: openai, azure-openai, ollama.
 *
 * Each provider implements EmbeddingProvider.embed(text) by calling the
 * provider's real HTTP embeddings API via fetch. The provider is chosen from
 * embedding.provider via createProvider().
 */

import type { EmbeddingConfig } from "./config-loader.js";

const AZURE_API_VERSION = "2024-02-01";

export interface EmbeddingProvider {
  embed(text: string): Promise<number[]>;
}

async function postJson(
  url: string,
  body: unknown,
  headers: Record<string, string>,
): Promise<any> {
  const response = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...headers },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    throw new Error(`Embedding request failed: ${response.status} ${response.statusText}`);
  }
  return response.json();
}

export class OpenAIProvider implements EmbeddingProvider {
  constructor(private readonly config: EmbeddingConfig) {}

  async embed(text: string): Promise<number[]> {
    const data = await postJson(
      "https://api.openai.com/v1/embeddings",
      { model: this.config.model, input: text },
      { Authorization: `Bearer ${this.config.apiKey}` },
    );
    return data.data[0].embedding;
  }
}

export class AzureOpenAIProvider implements EmbeddingProvider {
  constructor(private readonly config: EmbeddingConfig) {}

  async embed(text: string): Promise<number[]> {
    const endpoint = this.config.endpoint.replace(/\/+$/, "");
    const url =
      `${endpoint}/openai/deployments/${this.config.deploymentName}/embeddings` +
      `?api-version=${AZURE_API_VERSION}`;
    const data = await postJson(url, { input: text }, { "api-key": this.config.apiKey });
    return data.data[0].embedding;
  }
}

export class OllamaProvider implements EmbeddingProvider {
  constructor(private readonly config: EmbeddingConfig) {}

  async embed(text: string): Promise<number[]> {
    const endpoint = this.config.endpoint.replace(/\/+$/, "");
    const data = await postJson(
      `${endpoint}/api/embeddings`,
      { model: this.config.model, prompt: text },
      {},
    );
    return data.embedding;
  }
}

type ProviderCtor = new (config: EmbeddingConfig) => EmbeddingProvider;

const PROVIDER_REGISTRY: Record<string, ProviderCtor> = {
  openai: OpenAIProvider,
  "azure-openai": AzureOpenAIProvider,
  ollama: OllamaProvider,
};

export function createProvider(config: EmbeddingConfig): EmbeddingProvider {
  const ctor = PROVIDER_REGISTRY[config.provider];
  if (ctor === undefined) {
    throw new Error(
      `Unknown embedding provider: ${JSON.stringify(config.provider)}. ` +
        `Supported: ${Object.keys(PROVIDER_REGISTRY).sort().join(", ")}.`,
    );
  }
  return new ctor(config);
}
